pipeline_steps/common.groovy

```diff
@@ -27,10 +29,69 @@ def install_ansible(){
     pip install -c ${env.WORKSPACE}/rpc-gating/constraints.txt -U ansible pyrax
 
     mkdir -p rpc-gating/playbooks/roles
-    ansible-galaxy install -r ${env.WORKSPACE}/rpc-gating/role_requirements.yml -p ${env.WORKSPACE}/rpc-gating/playbooks/roles
+    ansible-galaxy install -r rpc-gating/role_requirements.yml -p rpc-gating/playbooks/roles
+
+    # The ansible version used on the Jenkins slave is always new enough for
+    # ara. So ARA will be used to collect data about playbooks run on the
+    # jenkins slave even if the ansible used by osa/rpco is too old. ARA is
+    # installed into the osa ansible runtime venv in branches after mitaka by
+    # rpco/scripts/deploy.sh
+    pip install -e ${env.ARA_REPO}@${env.ARA_BRANCH}#egg=ara
+    mkdir -p callbacks
+    cp .venv/src/ara/ara/plugins/callbacks/log_ara.py callbacks
   """
 }
 
+/* ARA records data about ansible task results,
+ * This function calls ara to generate a junit file from
+ * the ara database, and submits it to jenkins.
+ * This allows jenkins to record each ansible task as a
+ * unit test.
+ * Ara generate will dump all ansible task results from it's db into junit.xml,
+ * to prevent double reporting tests, junit.xml is cleared before and after
+ * submission to jenkins and the ara db is cleared afterwards.
+ */
+def submit_ara(){
+  dir(env.WORKSPACE){
+    db_file="ansible.sqlite"
+    if (fileExists(file: db_file)){
+      withEnv(common.get_deploy_script_env()){
+
+        // remove previously generated junit files and generate a new one.
+        sh """#!/bin/bash -x
+          . .venv/bin/activate
+          rm -f junit.xml ||:
+          ara generate junit junit.xml ||:
+        """
+
+        /* Submit results to jenkins.
+         * healthScaleFactor set to 0 so failed tests don't fail the build.
+         * There is no need for test results to contribute to build failure
+         * because the scripts that execute the tests will exit non 0 if there
+         * are failures and cause the build to fail that way.
+         * ARA reports failures with in a block/rescue as failed tests even
+         * if they are exepected failures, so if healthScaleFactor is not 0,
+         * these false negatives will cause all builds to fail.
+         * Tracking this in rcbops/u-suk-dev#1556
+         */
+        junit(
+          allowEmptyResults: true,
+          healthScaleFactor: 0,
+          testResults: 'junit.xml'
+        )
+
+        // remove submitted junit files and clear database
+        sh """#!/bin/bash -x
+          rm -f junit.xml ||:
+          rm -rf ${db_file}
+        """
+      }
+    } else {
+      print("submit_ara skipped as ${env.WORKSPACE}/${db_file} doesn't exist")
```

Comment:
Under what circumstances will this function be called when there is no ara db file?